### ISSUE TITLE
AWSIotMqttManager NoClassDefFoundError with Android < 7

### ISSUE THREAD

**Issue by oliveira-marcello** [aws-amplify/aws-sdk-android#1259]
In version 2.16.1, calling AWSIotMqttManager.connect causes NoClassDefFoundError below to be thrown. Version 2.16.0 doesn't have this problem. This is probably because 2.16.1 uses a newer paho version that causes this when Android API level is lower than 24 (https://github.com/eclipse/paho.mqtt.java/issues/633)
```
java.lang.NoClassDefFoundError: Failed resolution of: Ljavax/net/ssl/SNIHostName;
	at org.eclipse.paho.client.mqttv3.internal.SSLNetworkModule.start(SSLNetworkModule.java:139)
	at org.eclipse.paho.client.mqttv3.internal.ClientComms$ConnectBG.run(ClientComms.java:722)
	at java.lang.Thread.run(Thread.java:818)
 Caused by: java.lang.ClassNotFoundException: Didn't find class "javax.net.ssl.SNIHostName" on path: DexPathList[[zip file "/data/app/com.mercurio.ganimedes-1/base.apk"],nativeLibraryDirectories=[/data/app/com.mercurio.ganimedes-1/lib/arm, /vendor/lib, /system/lib]]
	at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:56)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:511)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:469)
	at org.eclipse.paho.client.mqttv3.internal.SSLNetworkModule.start(SSLNetworkModule.java:139) 
	at org.eclipse.paho.client.mqttv3.internal.ClientComms$ConnectBG.run(ClientComms.java:722) 
	at java.lang.Thread.run(Thread.java:818) 
	Suppressed: java.lang.ClassNotFoundException: javax.net.ssl.SNIHostName
	at java.lang.Class.classForName(Native Method)
	at java.lang.BootClassLoader.findClass(ClassLoader.java:781)
	at java.lang.BootClassLoader.loadClass(ClassLoader.java:841)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:504)
			... 4 more
 Caused by: java.lang.NoClassDefFoundError: Class not found using the boot class loader; no stack available
```

**To Reproduce**
Using Android with API level < 24
```
val mqtt = AWSIotMqttManager(clientId, CUSTOMER_SPECIFIC_ENDPOINT)
mqtt.keepAlive = 10
mqtt.isAutoReconnect = true
mqtt.setAutoResubscribe(true)

try {
	mqtt.connect(keyStore) { status, throwable ->
		Log.i(LOG_TAG, "AWS status = $status")
		if (throwable != null) {
			Log.e(LOG_TAG, "AWS IoT connection error", throwable)
		} else if (status == AWSIotMqttClientStatusCallback.AWSIotMqttClientStatus.Connected) {
			subscribe()
		}
	}
} catch (e: Exception) {
	Log.e(LOG_TAG, "AWS IoT connection error", e)
}
```

**Which AWS service(s) are affected?**
IoT

**Environment Information (please complete the following information):**
 - AWS Android SDK Version: 22
 - Device: Samsung Galaxy J1
 - Android Version: 5.1.1
 - Specific to simulators: no

**Comment by desokroshan** [comment]
@oliveira-marcello Thanks for reaching out. I will investigate and get back to you

**Comment by xanscale** [comment]
same problem here. various devices but all with android 6

lib ver: 2.16.1

**Comment by xanscale** [comment]
this changelog seams interesting:

> AWS Android SDK for IoT now depends on the 1.2.2 version of the Paho Library org.eclipse.paho.client.mqttv3

**Comment by xanscale** [comment]
https://developer.android.com/reference/javax/net/ssl/SNIHostName

this class is added in android 7.0

**Comment by xanscale** [comment]
https://github.com/eclipse/paho.mqtt.java/issues/633

**Comment by rynkowsg** [comment]
What is the status of this?

I'm developing an IoT app, and I'm not able to update the system. I'm stuck with API 22.

I don't want to use 2.16.0 forever when the lib is moving forward.
You have to either cut the support for older platforms officially or fix the issue either in your codebase or by contributing to Paho.

What's your plan?

**Comment by frittentheke** [comment]
@rynkowsg Eclipse Paho seriously lacks developers to fix the list of issues around race conditions, memory leaks and general stability - https://github.com/eclipse/paho.mqtt.java/issues

We moved to https://github.com/hivemq/hivemq-mqtt-client for this very reason.

**Comment by rynkowsg** [comment]
@frittentheke In the latest master you still seem to use Eclipse Paho. When and where did you move to hivemq-mqtt-client?

**Comment by frittentheke** [comment]
@rynkowsg ah sorry, I am not part of this project nor referring to it ... in another project which required an MQTT library we moved to the mentioned hivemq-mqtt-client.

**Comment by desokroshan** [comment]
@rynkowsg Sorry for the inconvenience. There already seems to be a [PR](https://github.com/Pitel/paho.mqtt.java/commit/27eebcb3b8a48c10d03f09e45eac6bd3cbf38ff2) with a suggested fix. The PR seems to make sense to me. Not sure how long will it take for paho team to review and release it though. We are discussing this internally whether to wait for a fix in Paho or to look for an alternative.  I do not have specific timeline that I can share at this point. I will let you know via an update on this thread as soon as there is any progress.

**Comment by rynkowsg** [comment]
@desokroshan I'm afraid that if we wait for Paho, we might never close this issue.

If you look at the Paho repository, it doesn't look like someone was coming to solve any issues there. 
The project is dying. The last commit is dated on 24 Sep 2019. The project's maintainer, Ian Craggs left IBM last year. He worked on Paho there. The Github user he used to use working on Paho no longer exist. He contributes now using a different account (@icraggs) but doesn't work on Java library.

January this year, he wrote a post http://modelbasedtesting.co.uk/2020/01/22/eclipse-paho-project-sponsorship-by-hivemq. Quotes:

> This is why I suggest that organizations which have an interest in the survival of Paho to consider encouraging their own people to become more deeply involved in the project.

> If there is insufficient interest to take actions to keep the project going then I am happy for the project or specific components, like the Java client, to be deprecated.

My reasoning is that the AWS SDK relies on an unmaintained/deprecated project. It is not good. Even for small projects, it is never good to use deprecated dependencies. Sooner or later it is going to be painful. Longer you wait, more painful it becomes.

This is AWS SDK. It is not a small project. All the more you should take action. Either by starting maintenance of the Paho Java library or by switching to an alternative that is maintained by 3rd party.

**Comment by raphkim** [comment]
Hi @rynkowsg ,

Thank you for your concerns. I will discuss with the team to see how we can approach this issue from our side.

**Comment by 103style** [comment]
Hi guys, i got the same bug too.

and i fixed it in my project  like the following setp:

* clone the  [eclipse/paho.mqtt.java](https://github.com/eclipse/paho.mqtt.java),

* catch SNIServerName not found bug under Android 7 with [#666](https://github.com/eclipse/paho.mqtt.java/pull/666)

* and use eclipse to make a new jar-- [org.eclipse.paho.client.mqttv3-1.2.2.jar](https://github.com/103style/paho.mqtt.java/blob/master/jar/org.eclipse.paho.client.mqttv3-1.2.2.jar)

* copy the jar  to libs in my project.

* replace the implementation in the build.gradle  like：
    ```
    implementation("com.amazonaws:aws-android-sdk-iot:$aws_version") {
        exclude module: 'org.eclipse.paho.client.mqttv3'
    }
    implementation files('libs/org.eclipse.paho.client.mqttv3-1.2.2.jar')
    ```

then it works under Android 7.

**Comment by raphkim** [comment]
Hi @103style ,

Thank you for providing a temporary workaround for those in need of immediate support. 

Fortunately, it appears that the [PR #666](https://github.com/eclipse/paho.mqtt.java/pull/666) has been merged by the Paho project maintainer. We will continue to track this issue to update our Paho dependency as soon as it is made available so that you do not need to manually import the SDK jar file.

**Comment by 103style** [comment]
@raphkim   Hi, `org.eclipse.paho.client.mqttv3 v1.2.3` is released.

**Comment by awc** [comment]
Why hasn't this problem been solved yet? After we updated the AWS SDK, there are many complaints about Android 7.0 crashes.

**Comment by raphkim** [comment]
Hi @awc ,

We are currently working on improving our testing infrastructure, which caused a small delay in delivering a fix for this issue. We are still looking to resolve it as quickly as possible, so please hang on tight!

**Comment by tom-domnic** [comment]
@raphkim Any update on this? As there are many users which are using Android lollipop and Marshmallow and reporting the same in very extensive level.

**Comment by tadas-subonis** [comment]
Another solution is to exclude a transitive paho dep and include a specific version manually:

```
    def aws_version = "2.16.+"
    implementation ("com.amazonaws:aws-android-sdk-iot:$aws_version"){
        exclude group: 'org.eclipse.paho', module: 'org.eclipse.paho.client.mqttv3'
    }
    implementation "org.eclipse.paho:org.eclipse.paho.client.mqttv3:1.2.3"
    implementation "com.amazonaws:aws-android-sdk-mobile-client:$aws_version"
```

**Comment by skyringtech** [comment]
Any update on this issue? Our new project on AWS IoT also needs to support Android 6 devices, and this bug is blocking our release. Thanks.

**Comment by raphkim** [comment]
This issue should have been resolved with the dependency updates from #1997 that made it into [v2.17.0](https://github.com/aws-amplify/aws-sdk-android/releases/tag/release_v2.17.0).
